app/feature/feature-travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/ui/overview/TravelCertificateOverviewViewModel.kt

```diff
@@ -0,0 +1,95 @@
+package com.hedvig.android.feature.travelcertificate.ui.overview
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import com.hedvig.android.feature.travelcertificate.data.DownloadTravelCertificateUseCase
+import com.hedvig.android.feature.travelcertificate.data.TravelCertificateUri
+import com.hedvig.android.feature.travelcertificate.data.TravelCertificateUrl
+import com.hedvig.android.logger.LogPriority
+import com.hedvig.android.logger.logcat
+import com.hedvig.android.molecule.android.MoleculeViewModel
+import com.hedvig.android.molecule.public.MoleculePresenter
+import com.hedvig.android.molecule.public.MoleculePresenterScope
+
+internal class TravelCertificateOverviewViewModel(
+  downloadTravelCertificateUseCase: DownloadTravelCertificateUseCase,
+) :
+  MoleculeViewModel<TravelCertificateOverviewEvent, TravelCertificateOverviewUiState>(
+      initialState = TravelCertificateOverviewUiState.Loading,
+      presenter = TravelCertificateOverviewPresenter(downloadTravelCertificateUseCase),
+    )
+
+internal class TravelCertificateOverviewPresenter(
+  private val downloadTravelCertificateUseCase: DownloadTravelCertificateUseCase,
+) : MoleculePresenter<TravelCertificateOverviewEvent, TravelCertificateOverviewUiState> {
+  @Composable
+  override fun MoleculePresenterScope<TravelCertificateOverviewEvent>.present(
+    lastState: TravelCertificateOverviewUiState,
+  ): TravelCertificateOverviewUiState {
+    var currentState by remember {
+      mutableStateOf<TravelCertificateOverviewUiState>(TravelCertificateOverviewUiState.Loading)
+    }
+
+    var dataLoadIteration by remember { mutableIntStateOf(0) }
+
+    var travelCertificateUrl by remember { mutableStateOf<TravelCertificateUrl?>(null) }
+
+    CollectEvents { event ->
+      when (event) {
+        TravelCertificateOverviewEvent.RetryLoadData -> {
+          dataLoadIteration++
+        }
+
+        is TravelCertificateOverviewEvent.OnDownloadCertificate -> {
+          travelCertificateUrl = event.travelCertificateUrl
+          dataLoadIteration++
+        }
+      }
+    }
+
+    LaunchedEffect(dataLoadIteration) {
+      val url = travelCertificateUrl
```

Comment:
```suggestion
    var travelCertificateUrl by remember { mutableStateOf<TravelCertificateUrl?>(null) }

    CollectEvents { event ->
      when (event) {
        TravelCertificateOverviewEvent.RetryLoadData -> {
          dataLoadIteration++
        }

        is TravelCertificateOverviewEvent.OnDownloadCertificate -> {
          travelCertificateUrl = event.travelCertificateUrl
        }
      }
    }

    LaunchedEffect(dataLoadIteration, travelCertificateUrl) {
      val url = travelCertificateUrl
```
Could this work here instead? So that the event handling still just does the simplest thing possible, and the rest of the presenter can figure out what that should do?

Now we are changing the `travelCertificateUrl` and then trigger the iteration as a side effect to run the LaunchedEffect down there because we are reading the value inside of there. Doing something like changing the order of setting the travelCertificateUrl and `dataLoadIteration++` might even break this functionality, and especially if we do more things in-between those two, so it feels a bit brittle this way.
If we instead key the `LaunchedEffect` also on the URL so that it re-runs the effect when it changes, we get the same behavior as far as I understand here.

Does this make sense? Am I perhaps missing something here?